cmd/util/ledger/migrations/cadence_values_migration.go

```diff
@@ -0,0 +1,243 @@
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/onflow/cadence/migrations"
+	"github.com/onflow/cadence/migrations/account_type"
+	"github.com/onflow/cadence/migrations/capcons"
+	"github.com/onflow/cadence/migrations/string_normalization"
+	"github.com/onflow/cadence/migrations/type_value"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+)
+
+type CadenceValueMigrator struct {
+	log      zerolog.Logger
+	reporter reporters.ReportWriter
+}
+
+var _ AccountBasedMigration = (*CadenceValueMigrator)(nil)
+var _ io.Closer = (*CadenceValueMigrator)(nil)
+
+func NewCadenceValueMigrator(
+	rwf reporters.ReportWriterFactory,
+) *CadenceValueMigrator {
+	return &CadenceValueMigrator{
+		reporter: rwf.ReportWriter("cadence-value-migrator"),
+	}
+}
+
+func (m *CadenceValueMigrator) Close() error {
+	// Close the report writer so it flushes to file.
+	m.reporter.Close()
+	return nil
+}
+
+func (m *CadenceValueMigrator) InitMigration(
+	log zerolog.Logger,
+	_ []*ledger.Payload,
+	_ int,
+) error {
+	m.log = log.With().Str("migration", "cadence-value-migration").Logger()
+	return nil
+}
+
+func (m *CadenceValueMigrator) MigrateAccount(
+	_ context.Context,
+	address common.Address,
+	oldPayloads []*ledger.Payload,
+) ([]*ledger.Payload, error) {
+
+	// Create all the runtime components we need for the migration
+	migrationRuntime, err := newMigratorRuntime(address, oldPayloads)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create migrator runtime: %w", err)
+	}
+
+	migration := migrations.NewStorageMigration(
+		migrationRuntime.Interpreter,
+		migrationRuntime.Storage,
+	)
+
+	capabilityIDs := map[interpreter.AddressPath]interpreter.UInt64Value{}
+
+	reporter := newValueMigrationReporter(m.reporter)
+
+	m.log.Info().Msg("Migrating cadence values")
+
+	migration.Migrate(
+		&migrations.AddressSliceIterator{
+			Addresses: []common.Address{
+				address,
+			},
+		},
+		migration.NewValueMigrationsPathMigrator(
+			reporter,
+			&capcons.CapabilityValueMigration{
+				CapabilityIDs: capabilityIDs,
+				Reporter:      reporter,
+			},
+			&capcons.LinkValueMigration{
+				CapabilityIDs:      capabilityIDs,
+				AccountIDGenerator: &AccountIDGenerator{},
+				Reporter:           reporter,
+			},
+			string_normalization.NewStringNormalizingMigration(),
+			account_type.NewAccountTypeMigration(),
+			type_value.NewTypeValueMigration(),
+		),
+	)
+
+	m.log.Info().Msg("Committing changes")
+	migration.Commit()
+
+	// finalize the transaction
+	result, err := migrationRuntime.TransactionState.FinalizeMainTransaction()
+	if err != nil {
+		return nil, fmt.Errorf("failed to finalize main transaction: %w", err)
+	}
+
+	// Merge the changes to the original payloads.
+	return m.mergeRegisterChanges(migrationRuntime, result.WriteSet)
+}
+
+func (m *CadenceValueMigrator) mergeRegisterChanges(
+	mr *migratorRuntime,
+	changes map[flow.RegisterID]flow.RegisterValue,
```

Comment:
`changes` might also contain completely new payloads.

Reply:
Good point. Added in fdb00687645d69802be39ec1779b76a8bac276d8